File: libs/ts/src/credentials/vault.ts
/**
 * Local vault — the encrypted-at-rest secret store (TypeScript port of the Rust reference).
 *
 * Single JSON file; AES-256-GCM records; envelope-wrapped DEK; HMAC over the canonical byte string.
 * Atomic write (tmp → rename) under a cross-process directory lock for the shared device vault;
 * reload-on-change reads; fail-closed on bad KEK/tamper. Not internally synchronized — the
 * {@link DefaultCredentialService} serializes access.
 */
import {
  existsSync,
  mkdirSync,
  readFileSync,
  renameSync,
  rmdirSync,
  statSync,
  writeFileSync,
} from "fs";
import { dirname } from "path";
import { randomUUID } from "crypto";

import { deriveMacKey, hmacSha256, hmacVerify, KEY_LEN, NONCE_LEN, open, random, seal } from "./crypto";
import { CredentialError } from "./errors";
import {
  FORMAT_VERSION,
  KekInfo,
  macInput,
  recordAad,
  SecretEntry,
  VaultFile,
  VersionEntry,
} from "./format";
import { KeyProvider } from "./keyprovider";
import { Secret, SecretMeta } from "./types";

const b64 = (b: Buffer): string => b.toString("base64");
const unb64 = (s: string): Buffer => Buffer.from(s, "base64");

export interface PutOptions {
  ttlSecs?: number;
  labels?: Record<string, string>;
  contentType?: string;
  source?: string;
  centralVersionId?: string;
}

export class LocalVault {
  private constructor(
    private readonly path: string,
    private readonly vaultId: string,
    private readonly dek: Buffer,
    // retained for phase-2 KEK rotation
    private readonly keyProvider: KeyProvider,
    private kek: KekInfo,
    private secrets: Record<string, SecretEntry>,
    private readonly keep: number,
    private stamp: string | null,
  ) {
    void this.keyProvider;
  }

  /**
   * Open (or create) the vault. `newVaultId`/`newDek` let a caller supply the vault id and DEK for a
   * brand-new vault (used by the async KMS key provider, which must pre-wrap the DEK out of band);
   * they are ignored for an existing vault and default to fresh random values.
   */
  static open(
    path: string,
    keyProvider: KeyProvider,
    keepVersions = 2,
    newVaultId?: string,
    newDek?: Buffer,
  ): LocalVault {
    const keep = Math.max(1, keepVersions);
    if (existsSync(path)) {
      const vf = readFile(path);
      if (vf.format !== FORMAT_VERSION) {
        throw new CredentialError(`unsupported vault format ${vf.format}`);
      }
      const dek = keyProvider.unwrapDek(vf.vaultId, vf.kek);
      verifyMac(dek, vf);
      return new LocalVault(path, vf.vaultId, dek, keyProvider, vf.kek, vf.secrets ?? {}, keep, fileStamp(path));
    }
    const dir = dirname(path);
    if (dir) {
      mkdirSync(dir, { recursive: true });
    }
    const vaultId = newVaultId ?? randomUUID();
    const dek = newDek ?? random(KEY_LEN);
    const kek = keyProvider.wrapDek(vaultId, dek);
    const v = new LocalVault(path, vaultId, dek, keyProvider, kek, {}, keep, null);
    v.save();
    return v;
  }

  get(name: string): Secret | undefined {
    const entry = this.secrets[name];
    if (!entry || entry.versions.length === 0) {
      return undefined;
    }
    return this.decrypt(name, entry.versions[entry.versions.length - 1]);
  }

  getVersion(name: string, version: string): Secret | undefined {
    const entry = this.secrets[name];
    if (!entry) {
      return undefined;
    }
    const v = entry.versions.find((x) => x.version === version);
    return v ? this.decrypt(name, v) : undefined;
  }

  exists(name: string): boolean {
    const entry = this.secrets[name];
    return !!entry && entry.versions.length > 0;
  }

  list(prefix = ""): SecretMeta[] {
    const names = Object.keys(this.secrets).sort((a, b) =>
      Buffer.compare(Buffer.from(a, "utf8"), Buffer.from(b, "utf8")),
    );
    const out: SecretMeta[] = [];
    for (const name of names) {
      if (!name.startsWith(prefix)) {
        continue;
      }
      const vs = this.secrets[name].versions;
      if (vs.length > 0) {
        out.push(metaOf(name, vs[vs.length - 1]));
      }
    }
    return out;
  }

  versions(name: string): string[] {
    return (this.secrets[name]?.versions ?? []).map((v) => v.version);
  }

  /** Upstream version id of the latest version of `name` (for sync change detection). */
  latestCentralVersionId(name: string): string | undefined {
    return this.secrets[name]?.versions.at(-1)?.centralVersionId;
  }

  put(name: string, plaintext: Buffer, opts: PutOptions = {}): string {
    const version = this.nextVersion(name);
    const nonce = random(NONCE_LEN);
    const ct = seal(this.dek, nonce, recordAad(this.vaultId, name, version), plaintext);
    const rec: VersionEntry = {
      version,
      createdMs: Date.now(),
      source: opts.source ?? "local",
      contentType: opts.contentType ?? "application/octet-stream",
      nonce: b64(nonce),
      ciphertext: b64(ct),
    };
    if (opts.ttlSecs !== undefined) {
      rec.ttlSecs = opts.ttlSecs;
    }
    if (opts.labels && Object.keys(opts.labels).length > 0) {
      rec.labels = { ...opts.labels };
    }
    if (opts.centralVersionId !== undefined) {
      rec.centralVersionId = opts.centralVersionId;
    }
    const entry = (this.secrets[name] ??= { versions: [] });
    entry.versions.push(rec);
    if (entry.versions.length > this.keep) {
      entry.versions.splice(0, entry.versions.length - this.keep);
    }
    this.save();
    return version;
  }

  delete(name: string): boolean {
    if (this.secrets[name]) {
      delete this.secrets[name];
      this.save();
      return true;
    }
    return false;
  }

  reloadIfChanged(): boolean {
    const cur = fileStamp(this.path);
    if (cur === this.stamp) {
      return false;
    }
    const vf = readFile(this.path);
    verifyMac(this.dek, vf);
    this.secrets = vf.secrets ?? {};
    this.kek = vf.kek;
    this.stamp = cur;
    return true;
  }

  private nextVersion(name: string): string {
    const last = this.secrets[name]?.versions.at(-1)?.version;
    const n = last ? Number.parseInt(last, 10) || 0 : 0;
    return String(n + 1).padStart(8, "0");
  }

  private decrypt(name: string, v: VersionEntry): Secret {
    const pt = open(this.dek, unb64(v.nonce), recordAad(this.vaultId, name, v.version), unb64(v.ciphertext));
    return new Secret(
      name,
      v.version,
      pt,
      v.labels ?? {},
      v.createdMs,
      v.source ?? "local",
      v.contentType ?? "application/octet-stream",
    );
  }

  private save(): void {
    const macKey = deriveMacKey(this.dek, this.vaultId);
    const mac = b64(hmacSha256(macKey, macInput(this.vaultId, this.secrets, unb64)));
    const vf: VaultFile = {
      format: FORMAT_VERSION,
      vaultId: this.vaultId,
      kek: this.kek,
      secrets: this.secrets,
      mac,
    };
    const data = JSON.stringify(vf, null, 2);
    withLock(`${this.path}.lock`, () => {
      const tmp = `${this.path}.tmp`;
      writeFileSync(tmp, data);
      renameSync(tmp, this.path);
    });
    this.stamp = fileStamp(this.path);
  }
}

function readFile(path: string): VaultFile {
  let raw: string;
  try {
    raw = readFileSync(path, "utf8");
  } catch (e) {
    throw new CredentialError(`read vault: ${(e as Error).message}`);
  }
  try {
    return JSON.parse(raw) as VaultFile;
  } catch (e) {
    throw new CredentialError(`parse vault: ${(e as Error).message}`);
  }
}

function verifyMac(dek: Buffer, vf: VaultFile): void {
  const macKey = deriveMacKey(dek, vf.vaultId);
  const expected = unb64(vf.mac);
  if (!hmacVerify(macKey, macInput(vf.vaultId, vf.secrets ?? {}, unb64), expected)) {
    throw new CredentialError("vault integrity check failed (tampered or wrong key)");
  }
}

function metaOf(name: string, v: VersionEntry): SecretMeta {
  return {
    name,
    version: v.version,
    createdMs: v.createdMs,
    ttlSecs: v.ttlSecs,
    source: v.source ?? "local",
    labels: v.labels ?? {},
  };
}

function fileStamp(path: string): string | null {
  try {
    const s = statSync(path);
    return `${s.mtimeMs}:${s.size}`;
  } catch {
    return null;
  }
}

/** Cross-process advisory lock via atomic `mkdir` (no native dep), with bounded retry. */
function withLock(lockDir: string, fn: () => void): void {
  const deadline = Date.now() + 5000;
  for (;;) {
    try {
      mkdirSync(lockDir);
      break;
    } catch {
      if (Date.now() > deadline) {
        throw new CredentialError("timed out acquiring vault lock");
      }
      const until = Date.now() + 15;
      while (Date.now() < until) {
        /* brief spin-wait; writes are tiny and rare */
      }
    }
  }
  try {
    fn();
  } finally {
    try {
      rmdirSync(lockDir);
    } catch {
      /* best effort */
    }
  }
}

File: libs/rust/src/credentials/vault.rs
//! # Local vault
//!
//! **One-liner purpose**: The encrypted-at-rest secret store — opens/creates the vault file,
//! seals/opens records under the DEK, persists atomically under a cross-process lock, and serves
//! versioned reads.
//!
//! ## Overview
//! A vault is a single JSON file ([`super::format::VaultFile`]). On open, the DEK is unwrapped via
//! the [`KeyProvider`] and the file MAC is verified; reads decrypt in memory. Writes append a new
//! monotonic version, prune to `keep_versions`, recompute the MAC, and persist with a temp→rename
//! atomic write while holding an advisory lock on a sidecar `.lock` file — so the **shared device
//! vault** stays consistent across concurrent component processes.
//!
//! ## Semantics & Architecture
//! - Single-writer-at-a-time via the file lock; readers are lock-free and see whole files only
//!   (atomic rename). [`LocalVault::reload_if_changed`] gives cross-process read freshness.
//! - All secret plaintext and key material live in [`zeroize`]-ing buffers.
//! - Errors map to [`GgError::Credentials`]; a MAC/AEAD failure is fail-closed (never returns
//!   partial/plaintext data).

use std::collections::BTreeMap;
use std::fs::File;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine as _;
use fs4::fs_std::FileExt;
use zeroize::Zeroizing;

use super::crypto;
use super::format::{self, KekInfo, SecretEntry, VaultFile, VersionEntry, FORMAT_VERSION};
use super::keyprovider::KeyProvider;
use super::service::{Secret, SecretMeta};
use crate::error::GgError;
use crate::Result;

/// Options controlling a `put`.
#[derive(Debug, Clone, Default)]
pub struct PutOptions {
    pub ttl_secs: Option<u64>,
    pub labels: BTreeMap<String, String>,
    pub content_type: Option<String>,
    /// `central` when written by the sync engine; defaults to `local`.
    pub source: Option<String>,
    pub central_version_id: Option<String>,
}

/// The encrypted local secret store.
pub struct LocalVault {
    path: PathBuf,
    vault_id: String,
    dek: Zeroizing<[u8; crypto::KEY_LEN]>,
    /// Retained past open for phase-2 KEK rotation / re-wrap (DEK lives in memory, so reads don't
    /// need it).
    #[allow(dead_code)]
    key_provider: Arc<dyn KeyProvider>,
    kek: KekInfo,
    secrets: BTreeMap<String, SecretEntry>,
    keep_versions: usize,
    /// (mtime, len) of the file as last loaded — drives [`reload_if_changed`].
    stamp: Option<(SystemTime, u64)>,
}

impl LocalVault {
    /// Open an existing vault or create a new empty one at `path`.
    ///
    /// # Post-conditions
    /// On success the DEK is unwrapped and (for an existing file) the MAC verified. A new vault is
    /// created with a random `vaultId` + DEK, the DEK wrapped via `key_provider`, and persisted.
    pub fn open(path: impl AsRef<Path>, key_provider: Arc<dyn KeyProvider>, keep_versions: usize) -> Result<Self> {
        let path = path.as_ref().to_path_buf();
        let keep_versions = keep_versions.max(1);
        if path.exists() {
            let vf = read_file(&path)?;
            Self::from_file(path, vf, key_provider, keep_versions)
        } else {
            if let Some(dir) = path.parent() {
                std::fs::create_dir_all(dir).map_err(|e| GgError::Credentials(format!("create vault dir: {e}")))?;
            }
            let vault_id = uuid::Uuid::new_v4().to_string();
            let dek = Zeroizing::new(crypto::random::<{ crypto::KEY_LEN }>());
            let kek = key_provider.wrap_dek(&vault_id, &dek)?;
            let mut v = Self {
                path,
                vault_id,
                dek,
                key_provider,
                kek,
                secrets: BTreeMap::new(),
                keep_versions,
                stamp: None,
            };
            v.save()?;
            Ok(v)
        }
    }

    fn from_file(path: PathBuf, vf: VaultFile, key_provider: Arc<dyn KeyProvider>, keep_versions: usize) -> Result<Self> {
        if vf.format != FORMAT_VERSION {
            return Err(GgError::Credentials(format!("unsupported vault format {}", vf.format)));
        }
        let dek = key_provider.unwrap_dek(&vf.vault_id, &vf.kek)?;
        verify_mac(&dek, &vf)?;
        let stamp = file_stamp(&path).ok();
        Ok(Self {
            path,
            vault_id: vf.vault_id,
            dek,
            key_provider,
            kek: vf.kek,
            secrets: vf.secrets,
            keep_versions,
            stamp,
        })
    }

    /// The vault's stable id.
    pub fn vault_id(&self) -> &str {
        &self.vault_id
    }

    /// Latest version of `name`, decrypted, or `None` if absent.
    pub fn get(&self, name: &str) -> Result<Option<Secret>> {
        match self.secrets.get(name).and_then(|e| e.versions.last()) {
            Some(v) => Ok(Some(self.decrypt(name, v)?)),
            None => Ok(None),
        }
    }

    /// A specific version of `name`, decrypted.
    pub fn get_version(&self, name: &str, version: &str) -> Result<Option<Secret>> {
        match self.secrets.get(name).and_then(|e| e.versions.iter().find(|v| v.version == version)) {
            Some(v) => Ok(Some(self.decrypt(name, v)?)),
            None => Ok(None),
        }
    }

    /// Whether a secret exists (no decryption).
    pub fn exists(&self, name: &str) -> bool {
        self.secrets.get(name).is_some_and(|e| !e.versions.is_empty())
    }

    /// Metadata for all secrets whose name starts with `prefix` (empty = all). Never decrypts.
    pub fn list(&self, prefix: &str) -> Vec<SecretMeta> {
        self.secrets
            .iter()
            .filter(|(n, _)| n.starts_with(prefix))
            .filter_map(|(n, e)| e.versions.last().map(|v| meta_of(n, v)))
            .collect()
    }

    /// Version ids retained for `name` (oldest→newest).
    pub fn versions(&self, name: &str) -> Vec<String> {
        self.secrets
            .get(name)
            .map(|e| e.versions.iter().map(|v| v.version.clone()).collect())
            .unwrap_or_default()
    }

    /// Write a new version of `name` and persist. Prunes to `keep_versions`.
    pub fn put(&mut self, name: &str, plaintext: &[u8], opts: PutOptions) -> Result<String> {
        let next = self.next_version(name);
        let nonce: [u8; crypto::NONCE_LEN] = crypto::random();
        let aad = format::record_aad(&self.vault_id, name, &next);
        let ct = crypto::seal(&self.dek, &nonce, &aad, plaintext)?;
        let entry = VersionEntry {
            version: next.clone(),
            created_ms: now_ms(),
            ttl_secs: opts.ttl_secs,
            source: opts.source.unwrap_or_else(|| "local".to_string()),
            central_version_id: opts.central_version_id,
            labels: opts.labels,
            content_type: opts.content_type.unwrap_or_else(|| "application/octet-stream".to_string()),
            nonce: B64.encode(nonce),
            ciphertext: B64.encode(ct),
        };
        let versions = &mut self.secrets.entry(name.to_string()).or_insert_with(|| SecretEntry { versions: vec![] }).versions;
        versions.push(entry);
        let keep = self.keep_versions;
        if versions.len() > keep {
            let drop = versions.len() - keep;
            versions.drain(0..drop);
        }
        self.save()?;
        Ok(next)
    }

    /// Remove a secret entirely and persist.
    pub fn delete(&mut self, name: &str) -> Result<bool> {
        let removed = self.secrets.remove(name).is_some();
        if removed {
            self.save()?;
        }
        Ok(removed)
    }

    /// Re-read the vault from disk if the file changed since last load (cross-process freshness).
    /// Returns `true` if a reload happened.
    pub fn reload_if_changed(&mut self) -> Result<bool> {
        let cur = file_stamp(&self.path).ok();
        if cur == self.stamp {
            return Ok(false);
        }
        let vf = read_file(&self.path)?;
        // DEK is unchanged for a given vault; re-verify integrity with the held DEK.
        verify_mac(&self.dek, &vf)?;
        self.secrets = vf.secrets;
        self.kek = vf.kek;
        self.stamp = cur;
        Ok(true)
    }

    fn next_version(&self, name: &str) -> String {
        let n = self
            .secrets
            .get(name)
            .and_then(|e| e.versions.last())
            .and_then(|v| v.version.parse::<u64>().ok())
            .unwrap_or(0);
        format!("{:08}", n + 1)
    }

    fn decrypt(&self, name: &str, v: &VersionEntry) -> Result<Secret> {
        let nonce: [u8; crypto::NONCE_LEN] = B64
            .decode(&v.nonce)
            .ok()
            .and_then(|b| b.try_into().ok())
            .ok_or_else(|| GgError::Credentials("bad record nonce".into()))?;
        let ct = B64.decode(&v.ciphertext).map_err(|_| GgError::Credentials("bad ciphertext".into()))?;
        let aad = format::record_aad(&self.vault_id, name, &v.version);
        let bytes = crypto::open(&self.dek, &nonce, &aad, &ct)?;
        Ok(Secret {
            name: name.to_string(),
            version: v.version.clone(),
            bytes: Zeroizing::new(bytes.to_vec()),
            labels: v.labels.clone(),
            created_ms: v.created_ms,
            source: v.source.clone(),
            content_type: v.content_type.clone(),
        })
    }

    /// Recompute the MAC and persist atomically under the cross-process lock.
    fn save(&mut self) -> Result<()> {
        let mac_key = crypto::derive_mac_key(&self.dek, &self.vault_id);
        let input = format::mac_input(&self.vault_id, &self.secrets, decode_b64);
        let mac = B64.encode(crypto::hmac(&mac_key, &input));
        let vf = VaultFile {
            format: FORMAT_VERSION,
            vault_id: self.vault_id.clone(),
            kek: self.kek.clone(),
            secrets: self.secrets.clone(),
            mac,
        };
        let json = serde_json::to_vec_pretty(&vf).map_err(|e| GgError::Credentials(format!("serialize vault: {e}")))?;

        // Coordinate concurrent writers (shared device vault) via an advisory lock on a sidecar.
        let lock_path = self.path.with_extension("lock");
        let lock = File::create(&lock_path).map_err(|e| GgError::Credentials(format!("open vault lock: {e}")))?;
        lock.lock_exclusive().map_err(|e| GgError::Credentials(format!("lock vault: {e}")))?;

        let tmp = self.path.with_extension("tmp");
        std::fs::write(&tmp, &json).map_err(|e| GgError::Credentials(format!("write vault tmp: {e}")))?;
        std::fs::rename(&tmp, &self.path).map_err(|e| GgError::Credentials(format!("rename vault: {e}")))?;
        let _ = FileExt::unlock(&lock);

        self.stamp = file_stamp(&self.path).ok();
        Ok(())
    }
}

fn read_file(path: &Path) -> Result<VaultFile> {
    let bytes = std::fs::read(path).map_err(|e| GgError::Credentials(format!("read vault: {e}")))?;
    serde_json::from_slice(&bytes).map_err(|e| GgError::Credentials(format!("parse vault: {e}")))
}

fn verify_mac(dek: &[u8; crypto::KEY_LEN], vf: &VaultFile) -> Result<()> {
    let mac_key = crypto::derive_mac_key(dek, &vf.vault_id);
    let input = format::mac_input(&vf.vault_id, &vf.secrets, decode_b64);
    let expected = B64.decode(&vf.mac).map_err(|_| GgError::Credentials("bad MAC encoding".into()))?;
    if crypto::hmac_verify(&mac_key, &input, &expected) {
        Ok(())
    } else {
        Err(GgError::Credentials("vault integrity check failed (tampered or wrong key)".into()))
    }
}

fn decode_b64(s: &str) -> Vec<u8> {
    B64.decode(s).unwrap_or_default()
}

fn meta_of(name: &str, v: &VersionEntry) -> SecretMeta {
    SecretMeta {
        name: name.to_string(),
        version: v.version.clone(),
        created_ms: v.created_ms,
        ttl_secs: v.ttl_secs,
        source: v.source.clone(),
        labels: v.labels.clone(),
    }
}

fn now_ms() -> u64 {
    SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_millis() as u64).unwrap_or(0)
}

fn file_stamp(path: &Path) -> std::io::Result<(SystemTime, u64)> {
    let m = std::fs::metadata(path)?;
    Ok((m.modified()?, m.len()))
}
